### PR TITLE
Jh/#1127 refactor multinomial logit

### PR DESCRIPTION
**Pull request by johanneshiry**
refactoring and generalization of the current implementation of the MultinomialLogit model to use it not only for ChooseMode but also for ChooseParking

---
This change is [Reviewable](https://reviewable.io/reviews/lbnl-ucb-sti/beam/1411)

### CODE REVIEW

File: src/main/scala/beam/agentsim/agents/choice/logit/MnlData.scala [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  * The parameters for each alternative that should be evaluated in the [[MultinomialLogit]] model
+  *
+  * @param alternativeId the general identifier of the alternative that should be evaluated (e.g. car)
+  * @param paramName the parameter identifier for the alternative (e.g. costs)
+  * @param paramType the type of the parameter as of [[UtilityParamType]] (e.g. multiplier)
+  * @param paramValue the value of the param type (e.g. -1.0)
+  * @tparam T
+  */
+case class MnlData[T](
+  alternativeId: T = "COMMON",
```
[the remark this answers is not in the document]
Comment (johanneshiry) [reply src/main/scala/beam/agentsim/agents/choice/logit/MnlData.scala]: I totally agree considering the "COMMON" parameter but I feel bad because I did pushed the changes of MnlData as I intended to drop this class totally (todo in line 17) because it is kinda duplicate of UtilityFunctionParams and could lead to confusion.

**Comment by johanneshiry** [comment]
test!

**Comment by JustinPihony** [comment]
Just for future reference, you can use the new Draft feature instead of requiring the WIP label

**Comment by johanneshiry** [comment]
test!

**Comment by johanneshiry** [comment]
test!

**Comment by johanneshiry** [comment]
test!

**Comment by johanneshiry** [comment]
test!

**Comment by johanneshiry** [comment]
test!

**Comment by johanneshiry** [comment]
test!

**Comment by johanneshiry** [comment]
test!

**Comment by JustinPihony** [comment]
Is this still a WIP? It's after April 15th, so it could be merged if everything was fixed up.

**Comment by robfitzgerald** [comment]
@JustinPihony this is still a WIP.

**Comment by johanneshiry** [comment]
test!

**Comment by johanneshiry** [comment]
test!

**Comment by johanneshiry** [comment]
test!

**Comment by johanneshiry** [comment]
test!

**Comment by johanneshiry** [comment]
test!

**Comment by johanneshiry** [comment]
test!

**Comment by johanneshiry** [comment]
test!

**Comment by johanneshiry** [comment]
test!

**Comment by johanneshiry** [comment]
test!

File: src/main/scala/beam/agentsim/agents/choice/mode/ModeChoiceLCCM.scala [hunk 2]
```diff
@@ -86,18 +87,18 @@ class ModeChoiceLCCM(
...
           val modeChoiceExpectedMaxUtility = lccm
             .modeChoiceModels(tourType)(theClassName)
-            .getExpectedMaximumUtility(modeChoiceInputData)
+            .getExpectedMaximumUtility(modeChoiceInputData.toMap)
           val surplusAttrib: Map[String, Double] =
-            Map("surplus" -> modeChoiceExpectedMaxUtility)
-          AlternativeAttributes(theClassName, attribIndivData.attributes ++ surplusAttrib)
-        }.toVector
+            Map("surplus" -> modeChoiceExpectedMaxUtility.getOrElse(0))
+          (theClassName, attribIndivData.head._2 ++ surplusAttrib)
```
Comment (JustinPihony) [inline src/main/scala/beam/agentsim/agents/choice/mode/ModeChoiceLCCM.scala]: I realize that this is hard-coded directly above and will currently never be empty, but we should add a note/test/something (empty check?) to make sure this is always the case.

File: src/main/scala/beam/agentsim/agents/choice/mode/ModeChoiceMultinomialLogit.scala [hunk 3]
```diff
@@ -257,48 +258,43 @@ class ModeChoiceMultinomialLogit(val beamServices: BeamServices, val model: Mult
...
-      new MnlData("bike", "intercept", "intercept", mnlConfig.params.bike_intercept),
-      new MnlData(
-        "walk_transit",
-        "intercept",
-        "intercept",
-        mnlConfig.params.walk_transit_intercept
+      "cav"       -> Map("intercept" -> UtilityFunctionOperation("intercept", mnlConfig.params.cav_intercept)),
+      "walk"      -> Map("intercept" -> UtilityFunctionOperation("intercept", mnlConfig.params.walk_intercept)),
+      "ride_hail" -> Map("intercept" -> UtilityFunctionOperation("intercept", mnlConfig.params.ride_hail_intercept)),
+      "walk"      -> Map("intercept" -> UtilityFunctionOperation("intercept", mnlConfig.params.walk_intercept)),
```
Comment (JustinPihony) [inline src/main/scala/beam/agentsim/agents/choice/mode/ModeChoiceMultinomialLogit.scala]: This is a duplicate, and I don't see anything that is missing from the original - so maybe just remove this?

File: src/main/scala/beam/agentsim/agents/choice/logit/MultinomialLogit.scala [hunk 4]
```diff
@@ -1,146 +1,170 @@
...
-                (theParam._2.paramValue * alternative.attributes(theParam._1)).toDouble
-              case Intercept =>
-                theParam._2.paramValue.toDouble
+  ): Option[MultinomialLogit.MNLSample[A]] = {
+    if (alternatives.isEmpty) None
+    else {
+
+      // one-pass evaluation and sorting in descending order by alternative utilities
+      val altsWithUtilitySortedDesc: Iterable[(A, Double)] =
+        alternatives.foldLeft(SortedSet.empty[(A, Double)](Ordering.by { -_._2 })) {
```
Comment (JustinPihony) [inline src/main/scala/beam/agentsim/agents/choice/logit/MultinomialLogit.scala]: We MAY want to be more explicit in our ordering as we've had some performance issues on the way generic ordering works. Thoughts @REASY?

**Comment by johanneshiry** [comment]
!test

**Comment by johanneshiry** [comment]
!test

**Comment by johanneshiry** [comment]
!test

**Comment by johanneshiry** [comment]
!test

**Comment by johanneshiry** [comment]
test!

**Comment by johanneshiry** [comment]
test!

**Comment by johanneshiry** [comment]
test!
